lib/sidekiq_rx_impl.cc

```diff
@@ -307,58 +370,109 @@ void sidekiq_rx_impl::apply_all_tags(size_t sample_index, size_t timestamp) {
 	tag_now = false;
 }
 
+
 int sidekiq_rx_impl::work(
 		int noutput_items,
 		gr_vector_const_void_star &,
-		gr_vector_void_star &output_items) {
-	unsigned int output_port{};
-	int samples_receive_count{};
+		gr_vector_void_star &output_items) 
+{
+	int samples_receive_count[MAX_PORT]{};
+	gr_complex *out[MAX_PORT];
+    static size_t last_timestamp_gap_update[MAX_PORT]{0,0} ;
 	unsigned int data_length_bytes{};
-	skiq_rx_block_t *p_rx_block{};
-	auto out = static_cast<gr_complex *>(output_items[output_port]);
-        static size_t last_timestamp_gap_update = 0;
 	int samples_to_rx = noutput_items*vector_length;
 	int num_vectors=0;
-	
-	if (nitems_written(output_port) - last_status_update_sample > status_update_rate_in_samples) {
-            if( timestamp_gap_count != last_timestamp_gap_update ) {
-		printf("Timestamp gap count: %ld\n", timestamp_gap_count);
-                last_timestamp_gap_update = timestamp_gap_count;
-            }
-		last_status_update_sample = nitems_written(output_port);
+	skiq_rx_block_t *p_rx_block{};
+    int port = 0;
+    int num_ports;
+    bool looping = false;
+    skiq_rx_hdl_t tmp_hdl;
+    int status = 0;
+   
```

Comment:
So in the old code, there was a variable called vector_length = DATA_MAX_BUFFER_SIZE.

The assumption was that the call to work is expecting samples_to_rx = noutput_items*vector_length

The call to work only has one variable of "noutput_items" passed in, even if there are two buffers in "output_items[]"

So the assumption is that every out buffer is the same size.

And in sidekiq_rx the buffer size is constant across the ports.

But, the code will continue to run until BOTH ports have received "samples_to_rx" number of samples.  

So this code will not return from work until both ports have filled up the buffers that the calling function sent in.

So if for some reason one port is getting more samples in than the other, this will continue waiting on skiq_receive to get more.

Is it possible for two ports at the same sample rate to get more samples on one port than the other?

I didn't think so.

Reply:
Yes, it is definitely possible with the current implementation for one handle to receive more data than the other (there are a handful of internal implementation details as to why this can occur).  So, I think we need to put some thought into how we'd like to handle this case and make sure we're handling this case.